Initial development

ESKYoung opened the pull request:
# Summary

Initial development leveraging [PyGithub][pygithub] to programmatically call GitHub's REST APIv3. The main aim is to prevent accidental loss of access to organisation repositories when an individual repository admin leaves the organisation. 

This code provides a way to find all organisation repositories, contributors, and teams, and to set wholesale permissions of a team to one or more repositories. 

Main functions include:

1. Getting all repositories from a GitHub organisation;
2. Executing PyGithub [`github.Repository.Repository` class methods][pygithub-repository] on one or more repositories;
3. Extracting PyGitHub object attributes from a list or [`github.PaginatedList.PaginatedList`][pygithub-pagination]; and
4. Adding a team to one of more repositories (if it does not already have access), whilst concurrently setting permissions levels.

Tests are written in pytest, and have full code coverage. Documentation has been added, including example Juypter notebooks in the `notebooks` folder.

# Review request

Grateful if the reviewer could check that the functions all work as expected (see the [comment below](https://github.com/ukgovdatascience/github_organisation_administration/pull/1#issuecomment-755274756) on adding and setting team permissions). Any comments on improvements that should be implemented would also be welcome.

I've tried to maintain security by enforcing use of a separate JSON file to store GitHub credentials (username and personal access token), but would be grateful for comments/suggestions!

# Checklists

<!--
These are DO-CONFIRM checklists; it CONFIRMs that you have DOne each item.

Outstanding actions should be completed before reviewers are assigned; if actions are irrelevant, please try and add a
comment stating why.

Incomplete pull/merge requests MAY be blocked until actions are resolved, or closed at the reviewers' discretion.
-->

This pull/merge request meets the following requirements:

- [x] Code runs
- [x] Developments are **secure** and [**ethical**][data-ethics-framework]
- [x] Developments follow the AQA plan (see `docs/aqa/aqa_plan.md`)
- [x] You have made proportionate checks that the code works correctly
- [x] Test suite passes
- [x] Assumptions, and caveats log updated (see `docs/aqa/assumptions_caveats.md`), if necessary
- [x] [Minimum usable documentation][agilemodeling] written in the `docs` folder, and in other Markdown files where
      relevant

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework
[pygithub]: https://pygithub.readthedocs.io
[pygithub-pagination]: https://pygithub.readthedocs.io/en/latest/utilities.html#pagination
[pygithub-repository]: https://pygithub.readthedocs.io/en/latest/github_objects/Repository.html

Commit:
Add parse_api_token function to parse a GitHub credentials JSON

This JSON should contain the GitHub username as well as the GitHub
personal access token. For further information about the token, see:
https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token

Commit:
Add code coverage Make command

Also ignore `if __name__ == "__main__":` sections from code coverage
reports.

Commit:
Add find_organisation_repos function including tests and documentation

Commit:
Add get_contributors_for_repo & get_contributors_for_all_repos functions

Including all tests and documentation.

Commit:
Refactor to split out get_contributors_for_repo function

Commit:
Refactor contributor functions into generic ones

Enables greater flexibility with GET requests on GitHub repositories.

Commit:
Add chunksize to multiprocessing functions

Can potentially speed up API requests even further.

Commit:
Split out parallel processing of dictionaries

Commit:
Add function to extract attribute from dictionary of PaginatedLists

Commit:
Add documentation, and rename project to better reflect purpose

Commit:
Unify if __name__ == "__main__" clauses for all functions

Commit:
Minor consolidation of test docstrings

Commit:
Add functions to add and set team to repository with permissions

Allows users to programmatically add a team to a repository (if not
already), and then set the permissions of that team. This can be done
at scale with multiple repositories.

Commit:
Refactor parallelisation and parallelise team permissions

Makes the setting team permissions for multiple repositories much
faster.

Commit:
Update function type hints to include lists as well as PaginatedLists

As long as the elements of the list are valid PyGithub objects, the
functions should work correctly.

Commit:
Add example Jupyter notebooks for various tasks

Also update documentation to reflect purpose of the project.

ESKYoung commented:
For clarity, I tested the adding and setting team permissions IRL by executing the `add_team_with_permissions_to_all_repositories` function on **just** the `govcookiecutter` repository. After ensuring my own GitHub username had admin rights (i.e. not through a team), I ran two tests:

1. Removing a team (if needed) from `govcookiecutter` manually, and then executing the function to give the team set privileges. Verifying the changes manually.
2. With an existing team on `govcookiecutter`, execute the function to give the team a different set of privileges. Verifying the changes manually.

Instructions on how to execute the function for single, named repository can be found in `notebooks/example_set_team_permission_for_all_organisation_repositories.ipynb` under the "Setting a team's permission for a specific repository name" subsection.

**I did not run this wholesale across the entire `ukgovdatascience` organisation** - predominately because it's dangerous without reviewing this code, and also because I am not an organisation owner so wouldn't have permissions to run this for many repositories.

chao-xian commented:
I haven't had a chance to read through the code/commits yet but just a thought on the JSON file holding credentials: It's not clear to me why we're not re-using the `.secrets` file to read in these values? I worry that these credentials would be accidentally committed in an unknown JSON file.

chao-xian commented on src/utils/parse_api_token.py:
```diff
@@ -0,0 +1,26 @@
+    """
+
+    # Read and parse the JSON file, and return only the key_username, and key_token keys in the file
+    with open(file, "r") as f:
```
It's not clear to me why we're not re-using the `.secrets` file to read in these values? I worry that these credentials would be accidentally committed in an unknown JSON file.

ESKYoung replied:
That's a very good point @chao-xian! I think by trying to add extra security, I've introduced a flaw! 😬 

I've made changes to the code in commit [`33db318`](https://github.com/ukgovdatascience/github_organisation_administration/pull/1/commits/33db31875d366a26a677c27fbd721e8eae295769) to remove the `parse_api_token`, and all its references, and revised the documentation to state that the API token should be added as an environment variable in the `.secrets` file. Let me know what you think!

chao-xian reviewed (commented):
I haven't had a chance to read through the code/commits yet but just a thought on the JSON file holding credentials

Commit:
Remove `parse_api_token` function, and all usage.

PR #1 comments revealed a potential security issue by placing the onus
on the user/contributor to place their API credentials JSON outside of
the repository. If they failed to do this, there is a risk they could
accidentially commit their credentials, especially if they have not
pre-installed pre-commit hooks.

To patch this security issue, `parse_api_token` and all its usages have
been removed. Instead the documentation now insists the user add their
API token to the `.secrets` file - this is an untracked file so should
prevent the credentials from being version-controlled.

PR #1 comment here:
https://github.com/ukgovdatascience/github_organisation_administration/pull/1#discussion_r552939555

Commit:
Refactor `GITHUB_API_TOKEN` to `GITHUB_API_KEY`

This ensures that `GITHUB_API_KEY` is picked up by the `detect-secrets`
pre-commit hook if it resides outside of the `.secrets` file.

oscarwyatt reviewed (approved):
Super work!
Had a close look and I don't have any comments. I guess the only thing I'd say is that if anything, your testing is probably _too_ thorough! To my mind, testing the parallelizing process probably isn't necessary, as it can be verified by someone developing it and it isn't tied up to the "business" logic - but I certainly wouldn't insist you remove it!

Sterling work :+1: :100:

ESKYoung commented:
Thanks for your review @oscarwyatt, and your helpful comments @chao-xian!
